File: python/224.py
from __future__ import print_function
xrange = range


class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []
        ans, op = 0, '+'
        i = 0
        while i < len(s):
            if s[i].isdigit():
                val = 0
                while i < len(s) and s[i].isdigit():
                    val = val * 10 + ord(s[i]) - ord('0')
                    i += 1
                ans = ans + val if op == '+' else ans - val
            elif s[i] in '+-':
                op = s[i]
                i += 1
            elif s[i] == '(':
                stack.append((ans, op))
                ans, op = 0, '+'
                i += 1
            elif s[i] == ')':
                pre, op = stack.pop()
                ans = pre + ans if op == '+' else pre - ans
                i += 1
            else:
                i += 1
        return ans

print(Solution().calculate("(1+(4+5+2)-3)+(6+8)"))

File: cpp/224.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int calculate(string s) {
    vector<int> vals; vector<char> ops;
    int n = s.size(), i = 0, ans = 0;
    char op = '+';
    while (i < n) {
      char c = s[i];
      if ('0' <= c && c <= '9') {
        int val = 0;
        while (i < n && '0' <= s[i] && s[i] <= '9') {
          val = val * 10 + s[i] - '0'; ++i;
        }
        ans = op == '+' ? ans + val : ans - val;
      } else if (c == '+' || c == '-') {
        op = c; ++i;
      } else if (c == '(') {
        vals.push_back(ans); ops.push_back(op);
        ans = 0; op = '+'; ++i;
      } else if (c == ')') {
        int pre = vals.back(); op = ops.back();
        vals.pop_back(); ops.pop_back();
        ans = op == '+' ? pre + ans : pre - ans;
        ++i;
      } else ++i;
    }
    return ans;
  }
};
